pydantic/tools.py

```diff
@@ -0,0 +1,20 @@
+from functools import lru_cache
+from typing import Any, Type, TypeVar
+
+__all__ = ('parse_as_type',)
+
+
+@lru_cache(maxsize=None)
+def _get_parsing_type(type_: Any, source: str) -> Any:
+    from pydantic.main import create_model
+
+    type_name = getattr(type_, '__name__', str(type_))
+    return create_model(f'ParsingModel[{type_name}] (for {source})', obj=(type_, ...))
```

Comment:
Not not sure about `obj` I think it could be confusing, I think we should use custom root types, e.g. `__root__`.

That might involve fixing #908 but I think that's a good idea anyway.